configuration/behind-proxy.md

```diff
@@ -0,0 +1,89 @@
+These will be different for your environment.
+Make sure to use the correct values for your proxy's FQDN, IP address, and ports instead of the examples provided in the below steps.
+
+~~~
+HTTP PROXY http://proxy.example.com:8080 or 10.0.0.1:8080
+HTTPS PROXY http://proxy.example.com:8443 or 10.0.0.1:8443 (in some proxies, this will be the same port as HTTP)
+FTP PROXY http://proxy.example.com:10021 or 10.0.0.1:10021 (this one too)
+~~~
+
+You can also use an authenticating proxy by adding a username and password in front, like:
+
+~~~
+HTTP PROXY http://username:password@proxy.example.com:8080 or username:password@10.0.0.1:8080
+HTTPS PROXY http://username:password@proxy.example.com:8443 or username:password@10.0.0.1:8443
+FTP PROXY http://username:password@proxy.example.com:10021 or username:password@10.0.0.1:10021
+~~~
+
+If you use special characters in your username or password, they need to be replaced with their `%` equivalent.
+See this [table](https://grox.net/utils/encoding.html).
+
+ 1. **Whonix**
+
+    If you are not using Whonix, skip to the next step.
+    
+    Configure `torrc` so that Whonix can connect through the proxy.
+    From the Qubes menu, go to `sys-whonix -> Tor User Config` and edit the `torrc` file to add these lines:
+    ~~~
+    DisableNetwork 0
+    HTTPproxy 10.0.0.1:8080
+    HTTPSproxy 10.0.0.1:8443
+    FascistFirewall 1
+    ~~~
+    It's important here to use the IP address instead of the proxy name.
+
+ 2. **Set apps to use proxy**
+
+    You can set `http_proxy` and `https_proxy` variables in your template(s) and all AppVMs based on them will automatically pick it up.
+    In Fedora based templates, create or edit `/etc/profile.d/proxy.sh`.
+    In Debian based templates, create or edit `/etc/environment` instead.
+    Then, add the following settings:
+    ~~~
+    export http_proxy=http://proxy.example.com:8080
+    export https_proxy=http://proxy.example.com:8443
+    export ftp_proxy=http://proxy.example.com:10021
+    export HTTP_PROXY=http://proxy.example.com:8080
+    export HTTPS_PROXY=http://proxy.example.com:8443
+    export FTP_PROXY=http://proxy.example.com:10021
+    ~~~
+    Here we use the fully qualified domain names instead of the proxy IP.
+
+ 3. **Allow UpdateVM to update via proxy**
+
+    If you receive system updates through Whonix, this step can be skipped and you are done!
+    If not, perform the following so updates can be downloaded through the proxy.
+
+    First, determine which template your UpdateVM is using:  
+     a. Go to a `dom0` terminal prompt.  
+     b. Enter `qubes-prefs` and look for `updatevm`.  
+     c. Enter `qvm-prefs <UpdateVMName>` and look for `template`.  
+    Perform the following steps from inside that template.
+
+    Add `proxy=http://10.0.0.1:8080` to the bottom of `/etc/apt/apt.conf.d/71proxy` in Debian templates.
+    In Fedora, add it to `/etc/dnf/dnf.conf` instead, after the line that says `### QUBES END ###`.
```

Comment:
This will not work in Qubes 4.0, because templates have no direct network access at all. Better to configure proxy in updates-proxy service. See `Upstream` setting in `tinyproxy.conf(5)`. This would require modifying `/etc/tinyproxy/tinyproxy-updates.conf`, including making it persistent (a file in `/rw` + `cp ... /etc/tinyproxy/tinyproxy-updates.con` in `/rw/config/rc.local`?).

Reply:
I thought adding the proxy setting here would result in it only getting used by the UpdateVM (e.g. sys-net) based on this template, and the templates themselves would actually continue to use qubes-rpc (4.0) or restricted networking (3.2) to pull updates from the UpdateVM. What am I missing?

Reply:
Thinking about it a bit more, I could see this proxy setting might override the one for `tinyproxy-updates` when using a debian-9 based sys-net to supply updates for the debian-9 template. Would it work to combine the approaches and apply the `apt.conf.d / dnf.conf` modifications strictly to the UpdateVM (instead of the underlying template) via your persistence approach?

Reply:
When you update template, dnf/apt is used only inside that template. And dnf/apt in template needs to be configured to use updates proxy - overriding proxy setting there will break updates. On the updates proxy side (sys-net) it is just http proxy, nothing more. Modifying dnf/apt configuration there have no effect on what is used by template to download updates.

There are some cases where modifying apt/dnf config to use some (corporate?) proxy in TemplateBasedVM is desirable - for example if you want to (temporarily) install package directly there, instead of installing it permanently in template. Another case is just checking for updates to decide whether template update is needed.
For those cases, it might be ok to set proxy setting _before_ `### QUBES BEGIN ###` (or in file ordered before 71proxy). This way, the setting will be used only in TemplateBasedVM, but will be overridden in TemplateVM to updates proxy. This needs to be tested, I'm not sure about apt and dnf behavior with duplicated settings.